SonarScanner.Shim/PropertiesFileGenerator.cs

```diff
@@ -102,48 +102,125 @@ public bool TryWriteProperties(PropertiesWriter writer, out IEnumerable<ProjectD
-                    .SelectMany(p => p.ExternalFiles)
-                    .Where(f => !PathHelper.IsPartOfAProject(f, projectPaths))
-                    .Distinct(StringComparer.InvariantCultureIgnoreCase)
-                    .ToList();
+                logger.LogError(Resources.ERR_NoValidProjectInfoFiles);
+                return false;
+            }
 
-                writer.WriteSonarProjectInfo(rootProjectBaseDir);
-                writer.WriteSharedFiles(sharedFilePaths);
+            writer.WriteSonarProjectInfo(rootProjectBaseDir);
+            writer.WriteSharedFiles(rootModuleFiles);
+            validProjects.ForEach(p => writer.WriteSettingsForProject(p, p.CoverageAnalysisExists(logger)
+                ? p.VisualStudioCoverageLocation : null));
 
-                foreach (var projectData in validProjects)
+            // Handle global settings
+            writer.WriteGlobalSettings(analysisProperties);
+
+            return true;
+        }
+
+        private IEnumerable<string> PutFilesToRightModuleOrRoot(IEnumerable<ProjectData> projects, string rootProjectBaseDir)
+        {
+            var fileWithProjects = projects
+                .SelectMany(p => p.ReferencedFiles.Select(f => new { Project = p, File = f }))
+                .GroupBy(group => group.File, StringComparer.OrdinalIgnoreCase)
+                .ToDictionary(group => group.Key, group => group.Select(x => x.Project).ToList());
+
+            foreach (var group in fileWithProjects)
+            {
+                var file = group.Key;
+
+                if (!File.Exists(file))
                 {
-                    writer.WriteSettingsForProject(projectData,
-                        projectData.CoverageAnalysisExists(logger) ? projectData.VisualStudioCoverageLocation : null);
+                    logger.LogWarning(Resources.WARN_FileDoesNotExist, file);
+                    logger.LogDebug(Resources.DEBUG_FileReferencedByProjects, string.Join("', '",
+                        group.Value.Select(x => x.Project.GetProjectDirectory())));
+                    continue;
                 }
 
-                // Handle global settings
-                writer.WriteGlobalSettings(analysisProperties);
+                if (!PathHelper.IsInFolder(file, rootProjectBaseDir)) // File is outside of the SonarQube root module
+                {
+                    logger.LogWarning(Resources.WARN_FileIsOutsideProjectDirectory, file);
+                    logger.LogDebug(Resources.DEBUG_FileReferencedByProjects, string.Join("', '",
+                        group.Value.Select(x => x.Project.GetProjectDirectory())));
+                    continue;
+                }
 
-                return true;
+                if (group.Value.Count >= 1)
+                {
+                    var closestProject = GetClosestProject(file, group.Value);
+
+                    if (closestProject == null)
+                    {
+                        yield return file;
```

Comment:
Using _yield_ here seems a bit odd. The method has side effects (updates _closestProject.SonarQubeModuleFiles_) so it isn't safe to suspend processing part way through. It would be better if the method built the complete list and then returned.